File: C/array/deque.c
/**
 * Author: Axel Ariel Saravia
 */

/*
    IMPORTANT

    the T type is the signed type

    deque struct is {
        T* content;
        size_t capacity;
        size_t head;
        size_t length;
    }
*/

# include "deque.h"

deque deque_create() {
    T* content = calloc(8, sizeof(T));
    return (
        !content
        ? (deque){0}
        : (deque){
            .content = content,
            .capacity = 8,
            .head = 0,
            .length = 0,
        }
    );
}

T deque_get(deque const dq, size_t const i) {
    if (i < dq.length) {
        size_t const idx = (i + dq.head) % dq.capacity;
        return dq.content[idx];
    }
    return 0;
}

T deque_set(deque dq[static 1], size_t const i, T const x) {
    if (i < dq->length) {
        size_t const idx = (i + dq->head) % dq->capacity; 
        T const y = dq->content[idx];
        dq->content[idx] = x;
        return y;
    }
    return 0;
}

deque deque_add(deque dq, size_t i, T const x) {
    size_t z1 = 0;
    size_t z2 = 0;
    if (dq.length < i) {
        i = dq.length;
    }
    if (dq.length == dq.capacity) {
        dq = deque_resize(dq);
    }
    if (i < dq.length / 2) {
        dq.head = (dq.head - 1) % dq.capacity;
        for (size_t k = 0; k < i; k += 1) {
            z1 = (dq.head + k) % dq.capacity;
            z2 = (dq.head + k + 1) % dq.capacity; 
            dq.content[z1] = dq.content[z2];
        }
    } else {
        for (size_t k = dq.length; k >= i + 1; k -= 1) {
            z1 = (dq.head + k) % dq.capacity;
            z2 = (dq.head + k - 1) % dq.capacity; 
            dq.content[z1] = dq.content[z2];
        }
    }
    z1 = (dq.head + i) % dq.capacity;
    dq.content[z1] = x;
    dq.length += 1;
    return dq;
}

deque deque_remove(deque dq, size_t i) {
    size_t z1 = (dq.head + i) % dq.capacity;
    size_t z2 = 0;
    if (dq.length <= i) {
        i = dq.length - i;
    }
    if (i < dq.length / 2) {
        for (size_t k = i; k; k -= 1) {
            z1 = (dq.head + k) % dq.capacity;
            z2 = (dq.head + k - 1) % dq.capacity; 
            dq.content[z1] = dq.content[z2];
        }
        dq.head = (dq.head + 1) % dq.capacity;
    } else {
        for (size_t k = i; k < dq.length - 1; k += 1) {
            z1 = (dq.head + k) % dq.capacity;
            z2 = (dq.head + k + 1) % dq.capacity; 
            dq.content[z1] = dq.content[z2];
        }
    }
    dq.length -= 1;

    if (8 < dq.capacity && 3 * dq.length <= dq.capacity) {
        dq = deque_resize(dq);
    }
    return dq;
}

deque deque_resize(deque dq) {
    size_t const length = dq.length;
    size_t const new_capacity = (
        8 < 2 * length
        ? 2 * length
        : 8
    );
    // this is never true when resize is called from add or remove methods
    if (new_capacity < length) {
        return dq;
    }
    T* new_content = 0;
    size_t const old_capacity = dq.capacity;
    size_t old_head = dq.head;
    size_t new_head = old_head;
    if (old_capacity < new_capacity) {

        do {
            new_content = realloc(dq.content, sizeof(T[new_capacity]));
        } while (!new_content);

        if (old_capacity < old_head + length) {
    // fstLength represents the first part of the dq
    // that is in the bottom of the array
            size_t const fst_length = old_capacity - old_head;
    // lstLength represents the last part of the dq
    // that is in the top of the array
            size_t const lst_length = length - fst_length;
            if (lst_length <= (new_capacity - old_capacity)) {
    // copy the values from the top of the array in to the bottom
                memcpy(
                    new_content + old_capacity,
                    new_content,
                    lst_length * sizeof(unsigned)
                );
            } else {
    // copy the values from the bottom of the array in to the top
                new_head = new_capacity - fst_length;
                memmove(
                    new_content + new_head,
                    new_content + old_head,
                    fst_length * sizeof(unsigned)
                );
            } 
        }
    // never can be newCapacity === oldCapacity
    } else {
        size_t fst_length = length;
        if (old_capacity < old_head + length) {
    // we must copy the bottom values at the end of the new array.
    // fstLength represent the first part of the dq
    //that is in the bottom of the array
            fst_length = old_capacity - old_head;
            new_head = new_capacity - fst_length;
        } else {
            new_head = 0;
        }
        memmove(
            dq.content + new_head,
            dq.content + old_head,
            fst_length * sizeof(unsigned)
        );
         do {
            new_content = realloc(dq.content, sizeof(T[new_capacity]));
        } while (!new_content);
    }
    dq.content = new_content;
    dq.capacity = new_capacity;
    dq.head = new_head;
    return dq;
}

void deque_free(deque* dq) {
    free(dq->content);
    *dq = (deque){0};
}

File: JavaScript/array/deque.js
/**
 * Author: Axel Ariel Saravia
 */

/**
Deque<T> :: {
    capacity: number[uint > 0],
    content: Array<T>,
    head: number[uint],
    length: number[uint]
}
*/

const ArrayDeque = Object.freeze({
    /**
    create :: () -> Deque<T> */
    create() {
        return Object.seal({
            capacity: 1,
            content: Array(1),
            head: 0,
            length: 0
        });
    },
    /**
    resize :: (Deque<T>) -> Deque<T> */
    resize(deque) {
        const length = deque.length;
        const newCapacity = Math.max(1, 2 * length);
    // this is never true when resize is called from add or remove methods
        if (newCapacity < length) {
            return deque;
        }
        const oldCapacity = deque.capacity;
        let oldHead = deque.head;
        let newHead = oldHead;
        if (oldCapacity < newCapacity) {
    // reallocate
            deque.content.length = newCapacity;
            if (oldCapacity < oldHead + length) {
    // fstLength represents the first part of the deque
    // that is in the bottom of the array
                const fstLength = oldCapacity - oldHead;
    // lstLength represents the last part of the deque
    // that is in the top of the array
                const lstLength = length - fstLength;
                if (lstLength <= (newCapacity - oldCapacity)) {
    // copy the values from the top of the array in to the bottom
                    for (let i = 0; i < lstLength; i += 1) {
                        deque.content[i + oldCapacity] = deque.content[i];
                    }
                } else {
    // copy the values from the bottom of the array in to the top
                    newHead = newCapacity - fstLength;
                    for (let i = 0; i < fstLength; i += 1) {
                        deque.content[i + newHead] = deque.content[i + oldHead];
                    }
                }
            }
    // never can be newCapacity === oldCapacity
        } else {
            let fstLength = length;
            if (oldCapacity < oldHead + length) {
    // we must copy the bottom values at the end of the new array.
    // fstLength represent the first part of the deque
    //that is in the bottom of the array
                fstLength = oldCapacity - oldHead;
                newHead = newCapacity - fstLength;
            } else {
                newHead = 0
            }
            for (let i = 0; i < fstLength; i += 1) {
                deque.content[i + newHead] = deque.content[i + oldHead];
            }
    // reallocate
            deque.content.length = newCapacity;
        }
        deque.head = newHead;
        deque.capacity = newCapacity;
        return deque;
    },
    /**
    get :: (Deque<T>, number[uint]) -> maybe<T> */
    get(deque, i) {
        if (0 <= i && i < deque.length) {
            const j = (i + deque.head) % deque.capacity;
            return deque.content[j];
        }
    },
    /**
    set :: (Deque<T>, number[uint], T) -> maybe<T> */
    set(deque, i, x) {
        if (0 <= i && i < deque.length) {
            const j = (i + deque.head) % deque.capacity;
            const y = deque.content[j];
            deque.content[j] = x;
            return y;
        }
    },
    /**
    add :: (Deque<T>, number[uint], T) -> Deque<T> */
    add(deque, i, x) {
        if (0 < i) {
            i = 0;
        } else if (deque.length < i) {
            i = deque.length;
        }
        if (deque.length === deque.capacity) {
            ArrayDeque.resize(deque);
        }
        if (i < deque.length / 2) {
            deque.head = (deque.head - 1) % deque.capacity;
            for (let k = 0; k < i - 1; k += 1) {
                const i0 = (deque.head + k) % deque.capacity;
                const i1 = (deque.head + k + 1) % deque.capacity;
                deque.content[i0] = deque[i1];
            }
        } else {
            for (let k = queue.length; k > i + 1; k -= 1) {
                const i0 = (deque.head + k) % deque.capacity;
                const i1 = (deque.head + k - 1) % deque.capacity;
                deque.content[i0] = deque[i1];
            }
        }
        const idx = (deque.head + i) % deque.capacity;
        deque.content[idx] = x;
        deque.length += 1;
        return queue;
    },
    /**
    remove :: (Deque<T>, number[uint]) -> T */
    remove(deque) {
        if (0 < i) {
            i = 0;
        } else if (deque.length <= i) {
            i = deque.length - 1;
        }
        const j = (deque.head + i) % deque.capacity;
        const x = deque.content[j];
        if (i < queue.length / 2) {
            for (let k = i; k > 0; k -= 1) {
                const i0 = (deque.head + k) % deque.capacity;
                const i1 = (deque.head + k - 1) % deque.capacity;
                deque.content[i0] = deque[i1];
            }
        } else {
            for (let k = i; k < deque.length; k += 1) {
                const i0 = (deque.head + k) % deque.capacity;
                const i1 = (deque.head + k + 1) % deque.capacity;
                deque.content[i0] = deque[i1];
            }
        }
        deque.length -= 1;
        if (3 * deque.length <= deque.capacity) {
            ArrayDeque.resize(deque);
        }
        return x;
    },
    /**
    clear :: (Deque<T>) -> undefined */
    clear(deque) {
        deque.capacity = 1;
        deque.content = Array(1);
        deque.head = 0;
        deque.length = 0;
    }
});

export default ArrayDeque;